constantine/ecdsa/ecdsa.nim

```diff
@@ -0,0 +1,356 @@
+import
+  ../hashes/h_sha256,
+  ../named/algebras,
+  ../math/io/[io_bigints, io_fields, io_ec],
+  ../math/elliptic/[ec_shortweierstrass_affine, ec_shortweierstrass_jacobian, ec_scalar_mul],
+  ../math/arithmetic,
+  ../platforms/abstractions,
+  ../serialization/codecs, # for fromHex and (in the future) base64 encoding
+  ../mac/mac_hmac, # for deterministic nonce generation via RFC 6979
+  ../named/zoo_generators, # for generator
+  ../csprngs/sysrand
+
+import std / macros # for `update` convenience helper
+
+type
+  ## Decides the type of sampler we use for the nonce. By default
+  ## a simple uniform random sampler. Alternatively a deterministic
+  ## sampler based on message hash and private key.
+  NonceSampler* = enum
+    nsRandom, ## pure uniform random sampling
+    nsRfc6979 ## deterministic according to RFC 6979
+
+# For easier readibility, define the curve and generator
+# as globals in this file
+const C* = Secp256k1
+const G = Secp256k1.getGenerator("G1")
+
+proc hashMessage(message: string): array[32, byte] =
+  # Hash a given message
+  var h {.noinit.}: sha256
+  h.init()
+  h.update(message)
+  h.finish(result)
+
+proc toBytes(x: Fr[C] | Fp[C]): array[32, byte] =
+  let bi = x.toBig()
+  discard result.marshal(bi, bigEndian)
+
+proc toDER*(r, s: Fr[C]): seq[byte] =
```

Comment:
There should be no allocations for signature algorithms.

Also DER should likely go in `constantine/serialization`

Reply:
Well, `toDER` is mainly for testing / convenience as also mentioned in its docstring. That's why I went with the simple `seq` approach here (mainly because we cannot determine at compile time what length the resulting DER encoded data has (due to prefix `0` byte or dropped leading zero bytes). Otherwise I assume one needs to always use an array that can fit the longest variant and truncate later.

Reply:
I would defer then to how libsecp256k1 (https://github.com/bitcoin-core/secp256k1) or BearSSL (https://bearssl.org/#download-and-installation) do it

libsecp256k1 is blockchain-focused so some TLS oriented serialization is not available.
Neither allocates.

But it's fine to put a note for now with TODO: zero-alloc

Reply:
Changed the implementation to something inspired by Bitcoin's implementation. Now have a `DERSignature` type  which has a 72 byte array buffer & an actual length and fill up to `len` bytes.

(We could of course also just have the proc take `sig: var array[72, byte], outlen: var int` parameters instead)